ledger-core/v2/virtual/request/vstaterequest.go

```diff
@@ -0,0 +1,194 @@
+
+type dSMVStateRequest struct {
+	smachine.StateMachineDeclTemplate
+}
+
+func (*dSMVStateRequest) InjectDependencies(sm smachine.StateMachine, _ smachine.SlotLink, injector *injector.DependencyInjector) {
+	s := sm.(*SMVStateRequest)
+
+	injector.MustInject(&s.pulseSlot)
+	injector.MustInject(&s.messageSender)
+}
+
+func (*dSMVStateRequest) GetInitStateFor(sm smachine.StateMachine) smachine.InitFunc {
+	s := sm.(*SMVStateRequest)
+	return s.Init
+}
+
+/* -------- Instance ------------- */
+
+func (*SMVStateRequest) InjectDependencies(sm smachine.StateMachine, _ smachine.SlotLink, injector *injector.DependencyInjector) {
+	s := sm.(*SMVStateRequest)
+
+	injector.MustInject(&s.pulseSlot)
+	injector.MustInject(&s.messageSender)
+}
+
+func (s *SMVStateRequest) GetStateMachineDeclaration() smachine.StateMachineDeclaration {
+	return dSMVStateRequestInstance
+}
+
+func (s *SMVStateRequest) Init(ctx smachine.InitializationContext) smachine.StateUpdate {
+	return ctx.Jump(s.stepProcess)
+}
+
+func (s *SMVStateRequest) stepProcess(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	var (
+		objectCatalog = object.Catalog{}
+	)
+	objectSharedState, stateFound := objectCatalog.TryGet(ctx, s.Payload.Callee)
+	if !stateFound {
+		s.failReason = payload.Missing
+		return ctx.Jump(s.stepReturnStateUnavailable)
+	}
+
+	var (
+		stateNotReady bool
+	)
+
+	action := func(state *object.SharedState) {
+		if !state.IsReady() {
+			stateNotReady = true
+			return
+		}
+
+		objectState := state.GetState()
+		switch objectState {
+		case object.Missing:
+			s.failReason = payload.Missing
+			return
+		case object.Inactive:
+			s.failReason = payload.Inactive
+			return
+		case object.HasState:
+		// ok case
+		default:
+			panic(throw.NotImplemented())
+		}
+
+		descriptor := state.Descriptor()
+
+		stateRef := descriptor.HeadRef()
+
+		s.objectStateReport = &payload.VStateReport{
+			AsOf:             s.Payload.AsOf,
+			Callee:           s.Payload.Callee,
+			LatestDirtyState: *stateRef,
+		}
+
+		if s.Payload.RequestedContent.Contains(payload.RequestLatestDirtyState) {
+			parent := descriptor.Parent()
+			proto, _ := descriptor.Prototype()
+
+			memory := descriptor.Memory()
+
+			s.objectStateReport.ProvidedContent = &payload.VStateReport_ProvidedContentBody{
+				LatestDirtyState: &payload.ObjectState{
+					Reference:   *stateRef,
+					Parent:      *parent,
+					State:       memory,
+					Deactivated: state.Deactivated,
+				},
+			}
+			if proto != nil {
+				s.objectStateReport.ProvidedContent.LatestDirtyState.Prototype = *proto
+			}
+		}
+	}
+
+	switch objectSharedState.Prepare(action).TryUse(ctx).GetDecision() {
+	case smachine.NotPassed:
+		return ctx.WaitShared(objectSharedState.SharedDataLink).ThenRepeat()
+	case smachine.Impossible:
+		ctx.Log().Fatal("failed to get object state: already dead")
+	case smachine.Passed:
+		// go further
+	default:
+		panic(throw.NotImplemented())
+	}
+
+	if stateNotReady {
```

Comment:
most of the time it will be stateNotReady, we set Ready to object only after getting VStateReport